File: src/main/java/io/curiousoft/ijudi/ordermanagement/model/StoreProfile.java
package io.curiousoft.ijudi.ordermanagement.model;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.util.StringUtils;

import javax.annotation.PostConstruct;
import javax.validation.Valid;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.PositiveOrZero;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.*;

public class StoreProfile extends Profile implements GeoPoint {

    @Indexed(unique = true)
    private String regNumber;
    @Indexed(unique = true)
    @NotBlank(message = "profile short name not valid")
    private String shortName;
    @NotBlank
    String ownerId;
    @Valid
    private Set<Stock> stockList = new HashSet<>();
    @NotEmpty(message = "profile tags not valid")
    private List<@NotEmpty String> tags;
    @NotEmpty(message = "Business hours not valid")
    private List<@NotNull BusinessHours> businessHours;
    private boolean hasVat;
    private boolean featured;
    private Date featuredExpiry;
    @NotNull(message = "storeType is not valid")
    private StoreType storeType;
    private Messager storeMessenger;
    private String storeWebsiteUrl;
    private boolean izingaTakesCommission;
    private boolean scheduledDeliveryAllowed = false;
    private AVAILABILITY availability = AVAILABILITY.SPECIFIC_HOURS;
    private String brandPrimaryColor = "#d69447";
    private String brandSecondaryColor = "#d69447";
    @PositiveOrZero(message = "free delivery min amount must be greater than or equal to 0.01")
    private double freeDeliveryMinAmount;
    private boolean markUpPrice = true;


    public StoreProfile(
                        @NotNull(message = "storeType is not valid") StoreType storeType,
                        @NotBlank(message = "profile name not valid") String name,
                        @NotBlank(message = "profile short name not valid") String shortName,
                        @NotBlank(message = "profile address not valid") String address,
                        @NotBlank(message = "profile image url not valid") String imageUrl,
                        @NotBlank(message = "profile mobile number not valid") String mobileNumber,
                        @NotEmpty(message = "profile tags not valid") List<String> tags,
                        @NotEmpty(message = "Business hours not valid") List<BusinessHours> businessHours,
                        @NotBlank(message = "shop owner id not valid") String ownerId,
                        @NotNull(message = "Shop bank not valid") Bank bank) {
        super(name, address, imageUrl, mobileNumber, ProfileRoles.STORE);
        this.businessHours = businessHours;
        this.tags = tags;
        this.ownerId = ownerId;
        this.storeType = storeType;
        this.shortName = shortName;
        setBank(bank);
    }

    public void setMarkUp(double markupPerc) {
        if(markUpPrice) {
            stockList.forEach(it -> it.setMarkupPercentage(markupPerc));
        }
    }

    public String getRegNumber() {
        return regNumber;
    }

    public void setRegNumber(String regNumber) {
        this.regNumber = regNumber;
    }

    public List<BusinessHours> getBusinessHours() {
        return businessHours;
    }

    public void setBusinessHours(List<BusinessHours> businessHours) {
        this.businessHours = businessHours;
    }

    public boolean getHasVat() {
        return hasVat;
    }

    public void setHasVat(boolean hasVat) {
        this.hasVat = hasVat;
    }

    public List<String> getTags() {
        return tags;
    }

    public void setTags(List<String> tags) {
        this.tags = tags;
    }

    public boolean getFeatured() {
        return featured;
    }

    public void setFeatured(boolean featured) {
        this.featured = featured;
    }

    public void setFeaturedExpiry(Date featuredExpiry) {
        this.featuredExpiry = featuredExpiry;
    }

    public Date getFeaturedExpiry() {
        return featuredExpiry;
    }

    public Set<Stock> getStockList() {
        return stockList;
    }

    public void setStockList(Set<Stock> stockList) {
        this.stockList = stockList;
    }

    public String getOwnerId() {
        return ownerId;
    }

    public void setOwnerId(String ownerId) {
        this.ownerId = ownerId;
    }

    public StoreType getStoreType() {
        return storeType;
    }

    public void setStoreType(StoreType storeType) {
        this.storeType = storeType;
    }

    public Messager getStoreMessenger() {
        return storeMessenger;
    }

    public void setStoreMessenger(Messager storeMessenger) {
        this.storeMessenger = storeMessenger;
    }

    public String getOrderUrl() {
        return !StringUtils.isEmpty(storeWebsiteUrl) ? storeWebsiteUrl + "/order/" : null;
    }

    public String getStoreWebsiteUrl() {
        return storeWebsiteUrl;
    }

    public void setStoreWebsiteUrl(String storeWebsiteUrl) {
        this.storeWebsiteUrl = storeWebsiteUrl;
    }

    public boolean getIzingaTakesCommission() {
        return izingaTakesCommission;
    }

    public void setIzingaTakesCommission(boolean izingaTakesCommission) {
        this.izingaTakesCommission = izingaTakesCommission;
    }

    public String getShortName() {
        return shortName;
    }

    public void setShortName(String shortName) {
        this.shortName = shortName != null ? shortName.toLowerCase() : null;
    }

    public boolean getScheduledDeliveryAllowed() {
        return scheduledDeliveryAllowed;
    }

    public void setScheduledDeliveryAllowed(boolean scheduledDeliveryAllowed) {
        this.scheduledDeliveryAllowed = scheduledDeliveryAllowed;
    }

    public String getBrandPrimaryColor() {
        return brandPrimaryColor;
    }

    public void setBrandPrimaryColor(String brandPrimaryColor) {
        this.brandPrimaryColor = brandPrimaryColor;
    }

    public String getBrandSecondaryColor() {
        return brandSecondaryColor;
    }

    public void setBrandSecondaryColor(String brandSecondaryColor) {
        this.brandSecondaryColor = brandSecondaryColor;
    }

    public double getFreeDeliveryMinAmount() {
        return freeDeliveryMinAmount;
    }

    public void setFreeDeliveryMinAmount(double freeDeliveryMinAmount) {
        this.freeDeliveryMinAmount = freeDeliveryMinAmount;
    }

    public boolean getMarkUpPrice() {
        return markUpPrice;
    }

    public void setMarkUpPrice(boolean markUpPrice) {
        this.markUpPrice = markUpPrice;
    }

    public AVAILABILITY getAvailability() {
        return availability;
    }

    public void setAvailability(AVAILABILITY availability) {
        this.availability = availability;
    }

    public boolean isEligibleForFreeDelivery(Order order) {
        return freeDeliveryMinAmount > 0  && order.getBasketAmount() >= freeDeliveryMinAmount;
    }

    public boolean isStoreOffline() {
        if(availability == AVAILABILITY.OFFLINE) {
            return true;
        }

        if(availability == AVAILABILITY.ONLINE24_7 || businessHours == null || businessHours.isEmpty()) {
            return false;
        }

        if(scheduledDeliveryAllowed) {
            return false;
        }

        Optional<BusinessHours> businessDay = businessHours.stream()
                .filter(day -> LocalDateTime.now().getDayOfWeek() == day.getDay())
                .findFirst();

        if(!businessDay.isPresent()) {
            return true;
        }

        Calendar calender = new Calendar.Builder().setInstant(businessDay.get().getOpen()).build();
        Date open  = Date.from(LocalDateTime.now()
                .withHour(calender.get(Calendar.HOUR_OF_DAY))
                .withMinute(calender.get(Calendar.MINUTE))
                .withSecond(calender.get(Calendar.SECOND))
                .atZone(ZoneId.systemDefault())
                .toInstant());

        calender = new Calendar.Builder().setInstant(businessDay.get().getClose()).build();
        Date close  = Date.from(LocalDateTime.now()
                .withHour(calender.get(Calendar.HOUR_OF_DAY))
                .withMinute(calender.get(Calendar.MINUTE))
                .withSecond(calender.get(Calendar.SECOND))
                .atZone(ZoneId.systemDefault())
                .toInstant());

        Date lowerBoundTime = Date.from(LocalDateTime.now().plusHours(2)
                .atZone(ZoneId.systemDefault())
                .toInstant());
        Date upperBoundTime = Date.from(LocalDateTime.now().plusHours(2).plusMinutes(14) //should not accept orders 15 minutes before store closes
                .atZone(ZoneId.systemDefault())
                .toInstant());
        return lowerBoundTime.before(open)
                || upperBoundTime.after(close);
    }

    public boolean isDeliverNowAllowed() {

        if(businessHours == null || businessHours.isEmpty()) {
            return false;
        }

        Optional<BusinessHours> businessDay = businessHours.stream()
                .filter(day -> LocalDateTime.now().getDayOfWeek() == day.getDay())
                .findFirst();

        if(!businessDay.isPresent()) {
            return false;
        }

        Calendar calender = new Calendar.Builder().setInstant(businessDay.get().getOpen()).build();
        Date open  = Date.from(LocalDateTime.now()
                .withHour(calender.get(Calendar.HOUR_OF_DAY))
                .withMinute(calender.get(Calendar.MINUTE))
                .withSecond(calender.get(Calendar.SECOND))
                .atZone(ZoneId.systemDefault())
                .toInstant());

        calender = new Calendar.Builder().setInstant(businessDay.get().getClose()).build();
        Date close  = Date.from(LocalDateTime.now()
                .withHour(calender.get(Calendar.HOUR_OF_DAY))
                .withMinute(calender.get(Calendar.MINUTE))
                .withSecond(calender.get(Calendar.SECOND))
                .atZone(ZoneId.systemDefault())
                .toInstant());

        Date lowerBoundTime = Date.from(LocalDateTime.now().plusHours(2)
                .atZone(ZoneId.systemDefault())
                .toInstant());
        Date upperBoundTime = Date.from(LocalDateTime.now().plusHours(2).plusMinutes(14) //should not accept orders 15 minutes before store closes
                .atZone(ZoneId.systemDefault())
                .toInstant());
        return lowerBoundTime.after(open)
                || upperBoundTime.before(close);
    }

    public enum AVAILABILITY {
        OFFLINE, SPECIFIC_HOURS, ONLINE24_7
    }
}

File: izinga-commons/src/main/kotlin/io/curiousoft/izinga/commons/model/StoreProfile.kt
package io.curiousoft.izinga.commons.model

import org.springframework.data.mongodb.core.index.Indexed
import org.springframework.util.StringUtils
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.*
import java.util.function.Consumer
import javax.validation.constraints.NotBlank
import javax.validation.constraints.NotEmpty
import javax.validation.constraints.NotNull
import kotlin.collections.HashSet

/**
 * A delivery or product category associated with a store.
 *
 * [name] must match a StockItem tag exactly (case-sensitive) — this invariant is enforced in
 * StoreService before any create/update that includes categories.
 *
 * [image] is intentionally left as an empty string or placeholder S3 URL at seed time;
 * the onboarding team uploads real images via the existing S3 document endpoint.
 * TODO #62: Verify S3 bucket ACL allows public-read on category images (DevOps/Lindani item — out of scope here).
 */
data class Category(
    val id: String,
    val name: String,
    val image: String,
    val active: Boolean
)

class StoreProfile(
    var storeType: @NotNull(message = "storeType is not valid") StoreType?,
    name: @NotBlank(message = "profile name not valid") String?,
    @field:Indexed(unique = true) private var shortName: @NotBlank(message = "profile short name not valid") String?,
    address: @NotBlank(message = "profile address not valid") String?,
    imageUrl: @NotBlank(message = "profile image url not valid") String?,
    mobileNumber: @NotBlank(message = "profile mobile number not valid") String?,
    var tags: @NotEmpty(message = "profile tags not valid") MutableList<String>?,
    var businessHours: @NotEmpty(message = "Business hours not valid") MutableList<BusinessHours>?,
    var ownerId: @NotBlank(message = "shop owner id not valid") String?,
    bank: @NotNull(message = "Shop bank not valid") Bank?
) : Profile(name, address, imageUrl, mobileNumber, ProfileRoles.STORE), GeoPoint {
    @Indexed(unique = true)
    var regNumber: String? = null

    var stockList: HashSet<Stock> = HashSet()
    var hasVat = false
    var featured = false
    var featuredExpiry: Date? = null
    var storeMessenger: HashSet<Messager>? = HashSet()
    var storeWebsiteUrl: String? = null
    var izingaTakesCommission = false
    var scheduledDeliveryAllowed = false
    var availability = AVAILABILITY.SPECIFIC_HOURS
    var freeDeliveryMinAmount = 0.0
    var markUpPrice = true
    var minimumDepositAllowedPerc = 1.0
    var standardDeliveryPrice = 0.0
    var standardDeliveryKm = 0.0
    var ratePerKm = 0.0
    var franchiseName: String? = null
    /** Delivery/product categories for this store. Never null — defaults to empty list. */
    var categories: List<Category> = emptyList()
    /** Whether the store has an active payment agreement with iZinga. */
    var hasPaymentAgreement: Boolean = false
    /** Whether the store can fulfill orders from more than one physical address. */
    var deliversFromMultipleAddresses: Boolean = false
    /** Whether the system should auto-generate missing stock images for this store. */
    var generateMissingImages: Boolean = false
    /** Whether a quote must be accepted by the customer before the order is confirmed. */
    var isQuoteRequired: Boolean = false
    /**
     * Per-vehicle-category delivery rate overrides, keyed by category label
     * (e.g. "Bike Delivery Driver", "Small/Medium Vehicle Driver", "Bakkie Delivery Driver", "Truck Delivery Driver").
     * When null the store-level [ratePerKm] is used; when present the matching key takes precedence.
     */
    var rates: MutableMap<String, Double>? = null

    init {
        super.bank = bank
    }

    fun setMarkUp(markupPerc: Double) {
        if (markUpPrice) {
            stockList.forEach(Consumer { it: Stock -> it.setMarkupPercentage(markupPerc) })
        }
    }

    val orderUrl: String?
        get() = if (!StringUtils.isEmpty(storeWebsiteUrl)) "$storeWebsiteUrl/order/" else null

    fun getShortName(): String? {
        return shortName
    }

    fun setShortName(shortName: String?) {
        this.shortName = shortName?.lowercase(Locale.getDefault())
    }

    fun isEligibleForFreeDelivery(order: Order): Boolean {
        return freeDeliveryMinAmount > 0 && order.basketAmount >= freeDeliveryMinAmount
    }

    //should not accept orders 15 minutes before store closes
    val isStoreOffline: Boolean
        get() {
            if (availability == AVAILABILITY.OFFLINE) {
                return true
            }
            if (availability == AVAILABILITY.ONLINE24_7 || businessHours == null || businessHours!!.isEmpty()) {
                return false
            }
            if (scheduledDeliveryAllowed) {
                return false
            }
            val businessDay = businessHours!!.stream()
                .filter { day: BusinessHours -> LocalDateTime.now().dayOfWeek == day.day }
                .findFirst()
            if (!businessDay.isPresent) {
                return true
            }
            var calender = Calendar.Builder().setInstant(businessDay.get().open).build()
            val open = Date.from(
                LocalDateTime.now()
                    .withHour(calender[Calendar.HOUR_OF_DAY])
                    .withMinute(calender[Calendar.MINUTE])
                    .withSecond(calender[Calendar.SECOND])
                    .atZone(ZoneId.systemDefault())
                    .toInstant()
            )
            calender = Calendar.Builder().setInstant(businessDay.get().close).build()
            val close = Date.from(
                LocalDateTime.now()
                    .withHour(calender[Calendar.HOUR_OF_DAY])
                    .withMinute(calender[Calendar.MINUTE])
                    .withSecond(calender[Calendar.SECOND])
                    .atZone(ZoneId.systemDefault())
                    .toInstant()
            )
            val lowerBoundTime = Date.from(
                LocalDateTime.now().plusHours(2)
                    .atZone(ZoneId.systemDefault())
                    .toInstant()
            )
            val upperBoundTime = Date.from(
                LocalDateTime.now().plusHours(2)
                    .plusMinutes(14) //should not accept orders 15 minutes before store closes
                    .atZone(ZoneId.systemDefault())
                    .toInstant()
            )
            return (lowerBoundTime.before(open)
                    || upperBoundTime.after(close))
        }

    //should not accept orders 15 minutes before store closes
    val isDeliverNowAllowed: Boolean
        get() {
            if (businessHours == null || businessHours!!.isEmpty()) {
                return false
            }
            val businessDay = businessHours!!.stream()
                .filter { day: BusinessHours -> LocalDateTime.now().dayOfWeek == day.day }
                .findFirst()
            if (!businessDay.isPresent) {
                return false
            }
            var calender = Calendar.Builder().setInstant(businessDay.get().open).build()
            val open = Date.from(
                LocalDateTime.now()
                    .withHour(calender[Calendar.HOUR_OF_DAY])
                    .withMinute(calender[Calendar.MINUTE])
                    .withSecond(calender[Calendar.SECOND])
                    .atZone(ZoneId.systemDefault())
                    .toInstant()
            )
            calender = Calendar.Builder().setInstant(businessDay.get().close).build()
            val close = Date.from(
                LocalDateTime.now()
                    .withHour(calender[Calendar.HOUR_OF_DAY])
                    .withMinute(calender[Calendar.MINUTE])
                    .withSecond(calender[Calendar.SECOND])
                    .atZone(ZoneId.systemDefault())
                    .toInstant()
            )
            val lowerBoundTime = Date.from(
                LocalDateTime.now().plusHours(2)
                    .atZone(ZoneId.systemDefault())
                    .toInstant()
            )
            val upperBoundTime = Date.from(
                LocalDateTime.now().plusHours(2)
                    .plusMinutes(14) //should not accept orders 15 minutes before store closes
                    .atZone(ZoneId.systemDefault())
                    .toInstant()
            )
            return (lowerBoundTime.after(open)
                    && upperBoundTime.before(close))
        }

    enum class AVAILABILITY {
        OFFLINE, SPECIFIC_HOURS, ONLINE24_7
    }
}